src/NuGet.Services.AzureSearch/VersionList/FilteredVersionList.cs

```diff
@@ -0,0 +1,203 @@
+                // the only listed version. The new latest version is of course the version we are adding right now.
+                Assert(ctx.NewLatest == ctx.ChangedVersion, "The first latest version must be the added version.");
+                Assert(ctx.NewProperties.Listed, "The added version should be listed for the latest version to have changed.");
+                return SearchIndexChangeType.AddFirst;
+            }
+
+            if (ctx.NewLatest == ctx.ChangedVersion)
+            {
+                // If the latest version has not changed or the latest version is the version we just added,
+                // then we need to update the existing metadata. This includes updating the latest version string.
+                return SearchIndexChangeType.UpdateExisting;
+            }
+
+            if (ctx.NewLatest < ctx.OldLatest)
+            {
+                // If the new latest is a lower version than the old latest, this is a special case where we need to
+                // look up the old new latest's metadata.
+                Assert(ctx.NewLatest != ctx.ChangedVersion, "This case should already have been handled.");
+                Assert(ctx.OldLatest == ctx.ChangedVersion, "This case should already have been handled.");
+                Assert(!ctx.NewProperties.Listed, "A downgrade from an upserted version can only happen from an unlist.");
+                return SearchIndexChangeType.DowngradeLatest;
+            }
+
+            // It's possible some cases (such as unlisting a version that didn't exist in the first place) could be
+            // handled without any change to the search index. However, to support the reflow case, we update the
+            // version list to make sure it is consistent.
+            return SearchIndexChangeType.UpdateVersionList;
+        }
+
+        private static void Assert(bool condition, string message)
+        {
+            /// We could use <see cref="System.Diagnostics.Debug.Assert(bool, string)"/> here, but it's preferable
+            /// in this case to even fail on a non-Debug build.
+            if (!condition)
+            {
+                throw new InvalidOperationException(message);
+            }
+        }
+
+        private Context UpdateVersionList(
+            Action<NuGetVersion, FilteredVersionProperties> update,
+            NuGetVersion version,
+            FilteredVersionProperties newProperties)
+        {
+            var oldLatest = _latest;
+
+            _versions.TryGetValue(version, out var oldProperties);
+
+            update(version, newProperties);
+
+            _latest = CalculateLatest();
+
+            return new Context(
+                version,
+                oldProperties,
+                newProperties,
+                oldLatest,
+                _latest);
+        }
+
+        private NuGetVersion CalculateLatest()
+        {
+            for (var i = _versions.Count - 1; i >= 0; i--)
+            {
```

Comment:
Could you use LINQ here? Maybe something like:

```
_versions.Reverse()
  .Where(v =>v.Value.Listed)
  .Select(v => v.Key)
  .FirstOrDefault();
```

Reply:
I don't see a big clarity win here and the performance is not obviously (to the reader) equivalent -- although it should be similar.

I don't care much so I'll change it.

Reply:
I agree it's not a huge difference, but I do think that `.Reverse()` is slightly easier to read than a manual for loop.

---
In reply to: [224177873](https://github.com/NuGet/NuGet.Services.Metadata/pull/368#discussion_r224177873) [](ancestors = 224177873)